File: python/searchable-client/src/searchable_client/synonyms.py
"""Synonym expansion helpers for the search client (mirrors
packages/client/src/synonyms.ts).
"""

from searchable_client.types import SynonymShard


def _synonym_variants_for(term: str, synonym_shard: SynonymShard | None) -> list[str]:
    if synonym_shard is None:
        return []
    variants: set[str] = set()
    for group in synonym_shard.equivalences:
        if term in group:
            variants.update(v for v in group if v != term)
    variants.update(synonym_shard.directional.get(term, []))
    return list(variants)


def _multi_word_variants_for(
    normalized_phrase: str, synonym_shard: SynonymShard | None
) -> list[str]:
    """Every other normalized phrase `normalized_phrase` expands to via the
    synonym shard's `multiWord` equivalence classes (docs/guides/synonyms.md).
    Symmetric only, matching the TS `multiWordVariantsFor`."""
    if synonym_shard is None:
        return []
    variants: set[str] = set()
    for group in synonym_shard.multi_word:
        if normalized_phrase in group:
            variants.update(v for v in group if v != normalized_phrase)
    return list(variants)

File: packages/searchable/src/synonyms.ts
import { ownProp } from "./analysis/index.js";
import type { SynonymShard } from "./format/index.js";

/** Every other term `term` expands to via the synonym shard's equivalence classes and directional map (docs/guides/synonyms.md). */
export function synonymVariantsFor(
  term: string,
  synonymShard: SynonymShard | undefined,
): string[] {
  if (!synonymShard) return [];
  const variants = new Set<string>();
  for (const group of synonymShard.equivalences ?? []) {
    if (!group.includes(term)) continue;
    for (const variant of group) {
      if (variant !== term) variants.add(variant);
    }
  }
  const directionalVariants = synonymShard.directional
    ? ownProp(synonymShard.directional, term)
    : undefined;
  for (const variant of directionalVariants ?? []) {
    variants.add(variant);
  }
  return [...variants];
}

/**
 * Every other normalized phrase `normalizedPhrase` expands to via the
 * synonym shard's `multiWord` equivalence classes
 * (docs/guides/synonyms.md#synonym-file-format) — symmetric only, no
 * directional multiWord form is defined. `normalizedPhrase` must
 * already be in the same space-joined-analyzed-terms shape
 * `normalizePhrase()`/the indexer's `buildSynonymShards()` produce.
 */
export function multiWordVariantsFor(
  normalizedPhrase: string,
  synonymShard: SynonymShard | undefined,
): string[] {
  if (!synonymShard) return [];
  const variants = new Set<string>();
  for (const group of synonymShard.multiWord ?? []) {
    if (!group.includes(normalizedPhrase)) continue;
    for (const variant of group) {
      if (variant !== normalizedPhrase) variants.add(variant);
    }
  }
  return [...variants];
}
